builtin/providers/postgresql/resource_postgresql_role_funcs.go

```diff
@@ -0,0 +1,113 @@
+package postgresql
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/lib/pq"
+)
+
+func resourcePostgresqlRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*sql.DB)
+	roleName := d.Get("name").(string)
+	loginAttr := getLoginStr(d.Get("login").(bool))
+	password := d.Get("password").(string)
+
+	encryptedCfg := getEncryptedStr(d.Get("encrypted").(bool))
+
+	query := fmt.Sprintf("CREATE ROLE %s %s %s PASSWORD '%s'", pq.QuoteIdentifier(roleName), loginAttr, encryptedCfg, password)
+	_, err := conn.Query(query)
+	if err != nil {
+		return fmt.Errorf("Error creating role: %s", err)
+	}
+
+	d.SetId(roleName)
+
+	return nil
+}
+
+func resourcePostgresqlRoleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*sql.DB)
+	roleName := d.Get("name").(string)
+
+	query := fmt.Sprintf("DROP ROLE %s", pq.QuoteIdentifier(roleName))
+	_, err := conn.Query(query)
+	if err != nil {
+		return err
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourcePostgresqlRoleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*sql.DB)
+	roleName := d.Get("name").(string)
+
+	var canLogin bool
+	err := conn.QueryRow("select rolcanlogin from pg_roles where rolname=$1", roleName).Scan(&canLogin)
+	switch {
+	case err == sql.ErrNoRows:
+		d.SetId("")
+		return nil
+	case err != nil:
+		return fmt.Errorf("Error reading info about role: %s", err)
+	default:
+		d.Set("login", canLogin)
+		return nil
+	}
+}
+
+func resourcePostgresqlRoleUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*sql.DB)
+	roleName := d.Get("name").(string)
+
+	if d.HasChange("login") {
+		loginAttr := getLoginStr(d.Get("login").(bool))
+		query := fmt.Sprintf("ALTER ROLE %s %s", pq.QuoteIdentifier(roleName), pq.QuoteIdentifier(loginAttr))
+		_, err := conn.Query(query)
+		if err != nil {
+			return fmt.Errorf("Error updating login attribute for role: %s", err)
+		}
+	}
+
+	password := d.Get("password").(string)
+
+	if d.HasChange("password") {
+		encryptedCfg := getEncryptedStr(d.Get("encrypted").(bool))
+
+		query := fmt.Sprintf("ALTER ROLE %s %s PASSWORD '%s'", pq.QuoteIdentifier(roleName), encryptedCfg, password)
+		_, err := conn.Query(query)
```

Comment:
In cases like this where you need to do multiple calls to update different parts of the object, and each one of these calls can separately fail, it's good to use a `ResourceData` mechanism called *partial state*, which in the event of an error allows Terraform to understand which values have been updated and which haven't, and thus reflect the current values in the state.

The pattern for partial state works like this:
* At the top of this function, before the first `HasChange` test, enable partial state mode by calling `d.Partial(true)`.
* After each one of these conditional calls succeeds, record the partial update by calling (for example) `d.SetPartial("password")`.
* Just before the `return` statement at the very end of the function, call `d.Partial(false)` to disable partial mode.